File: python/basics/hash_table/1_two_sum.py
'''
https://leetcode.com/problems/two-sum/description/?envType=study-plan-v2&envId=top-interview-150

#O(n)
hashtable[] forward lookup

#meta
'''

def two_sum(nums, target):
    result = []
    hash = {}
    n = len(nums)
    for i in range(n):
        remain = target - nums[i]
        if remain in hash:
            return [i, hash[remain]]
        else:
            hash[nums[i]] = i

    return result


nums = [2,7,11,15]
target = 9
result = two_sum(nums, target)
print("result: " + str(result))

nums = [3,2,4]
target = 6
result = two_sum(nums, target)
print("result: " + str(result))


nums = [3,3]
target = 6
result = two_sum(nums, target)
print("result: " + str(result))

File: problem_solving/hash_table/cpp/1_two_sum.cpp
#include <vector>
#include <map>
#include "../../utils/utils.h"

std::vector<int> two_sum(const std::vector<int>& nums, int target) {
    std::map<int, int> hashmap;
    int len = nums.size();
    for(auto i = 0; i < len; ++i) {
        if (hashmap.find(target - nums[i]) != hashmap.end()) {
            return {hashmap[target-nums[i]], i};
        } 
        hashmap[nums[i]] = i;
    }
    return {};
}


//print_vector(result);
//Output: [0,1]



/*

*/

int main() {
    std::vector<int> nums = {2,7,11,15};
    int target = 9;
    auto result = two_sum(nums, target);
    print_vector(result);

    nums = {3,2,4};
    target = 6;
    result = two_sum(nums, target);
    print_vector(result);
    //utput: [1,2]


    nums = {3,3};
    target = 6;
    result = two_sum(nums, target);
    print_vector(result);
    //Output: [0,1]

}